### DISCUSSION TITLE
Error Recovery

### DISCUSSION THREAD

**Discussion by DavidSouther** [ohmjs/ohm#492] [Ideas]
As I understand it, today, Ohm's parser will stop on the first syntax error. What would it look like to have error recovery available and implemented? I see a couple issues and discussions, namely

https://github.com/ohmjs/ohm/issues/387
https://github.com/ohmjs/ohm/issues/302
https://github.com/ohmjs/ohm/issues/265

**Comment by pdubroy** [comment]
I definitely find the labeled failures approach interesting! I'd be happy to support you (or anyone else) who's interested in digging in to try to prototype this.

**Reply by DavidSouther** [reply]
I dug at it a couple years ago, but I can't find my notes from the time. I dug at it again more recently, and came up with `!` as a "Fallible" marker, simplifying the Labeled Failure approach a bit. The idea being `A B! C` would allow B to fail with an error. In practice, I could see a language using this like `statement = expression ! ';'` - a statement tries to read an expression, but if it can't, it "resets" at the next `;`. This essentially works as a non-greedy `*` and captures that information into an "Error" semantic type.

I don't remember where this ended up in practice, but here's the diff: https://github.com/DavidSouther/ohm/pull/1

(I also have this change to streamline testing using the editor, it looks like - https://github.com/DavidSouther/ohm-editor/pull/1)

I might have some time in the next couple weeks to pick this up again, but no promises!
